Site editor: integrate global styles controls and style book preview into the styles panel

jorgefilipecosta opened the pull request:
Part of: https://github.com/WordPress/gutenberg/issues/53483

Closes #66376

This PR is a first step towards this vision proposed by @jameskoster and @jasmussen:

https://github.com/user-attachments/assets/4de07f95-37da-43a0-892b-51754ef9ac8a


Global styles controls are now shown in the style panel. The default preview is "Site", but users can activate the style book as well.


## Screenshots

This PR:


https://github.com/user-attachments/assets/d165f443-412f-46ce-a8db-e16ddcfb9b56





## Next steps
- Add styles section to classic themes (read-only style book in that case)
- Implement the improved overview style book shown in the video.
- Navigate to different sections of the style book depending on if the user is on typography, colors or a block.

jorgefilipecosta committed:
Update: Make the styles sections a preview of the style book with a left edit styles sidebar.

jorgefilipecosta committed:
improve mobile behavior

youknowriad commented:
Great start. I feel this gives already a good sense of pros and cons.

Here's some remarks about the experience.

 - I really like that I'm only focusing on on style changes on this page.
 - All edits of global styles are properly reflected in the stylebook, except the "layout" changes. I think there's a few things that we can do here:
     - We can have the "content width" apply automatically to all blocks in the style book
     - We can have a dedicated section of the style book that is specific to "layout", so when you navigate to "layout", the stylebook also navigates to it and we show for instance multiple blocks: "wide", "full" and "default" aligned and see how the changes impact them. Also block spacing should impact them there as well.
 - The header area of the stylebook tab is kind of missing. It would be good to bring these features somewhere :) (style revisions, custom CSS, revert...)
 
<img width="267" alt="Screenshot 2024-09-25 at 9 47 02 AM" src="https://github.com/user-attachments/assets/df06c73f-3603-411f-9271-17dc113eb3c2">

 But globally, I think this feels way better than what we have on trunk. We should try to gather more feedback here from everyone @WordPress/gutenberg-design

jasmussen commented:
Nice work. There are some strong aspects to this, mainly in that it places _global changes_ in context of your _site overall_, rather than in context of editing an individual page or template as is the case now. This, in turn, can also benefit the template editing experience where sometimes you'll confuse the block inspector with the global styles inspector.

The white second panel takes up a lot of space, I like to think there are some longer term things we can do here, possibly even collapse the dark material on the left, or show some or all of the style pieces in the dark material (potentially pending #56669). But that's a future thing, and the white second column at the moment makes it possible to make good progress on this with speed.

Revisions, custom CSS, revert, these are pieces we need to port over in the short term, but also elevate better in the longer term. Notably revisions is something that can potentially benefit from being thought of as part of the save dialog, or just the save button itself. CC: @auareyou as I know you've looked at the save dialog.

One thing we need to do is decide what clicking on items in the frame does. Does it open the fullscreen editor? Or does it select individual block types? If it's the latter, then it should navigate the the appropriate block in the 2nd column, just like it does in trunk. 

We need the inverse too, perhaps not in this branch but eventually: when you click "Blocks → Image", we should load only the image block preview in the frame on the right. This is going to especially benefit [interfaces as those proposed here](https://github.com/WordPress/gutenberg/issues/38277#issuecomment-2306449475), where the full range of block states can be shown: hover, active, focus, etc.

This way of making the preview frame contextual to what is selected in the second column can also potentially be leveraged to benefit what's being outlined in #53431:

* The "Styles" section loads the new curated overview section
* Individual blocks (Blocks → Block name) opens the individual block in the preview
* Typography opens [the Text list](https://www.figma.com/design/guZBzJKVdWLcNeWCZ7Rvx1/Style-Book?node-id=0-928&t=GRkHwhfoZ4lsoD3c-1)
* Colors opens [the Color list](https://www.figma.com/design/guZBzJKVdWLcNeWCZ7Rvx1/Style-Book?node-id=0-2301&t=GRkHwhfoZ4lsoD3c-1)
* Shadows opens a to-be-designed exploded view of shadow presets

But a first step might be to simply load a particular block in the 2nd column, when it is selected in the interface on the right.

Also, should we make this an experiment so it can be tested easily?

youknowriad commented:
> Also, should we make this an experiment so it can be tested easily?

I think we should make this an experiment if we judge that the proposal is a temporarily worse experience than trunk. To be honest, I'm not entirely sure :P I feel like this is already better than trunk. But I also understand if you want to keep it "more hidden" until we iron out more things.

youknowriad commented:
For me, we could make the following update as well:

 - Leave the dark sidebar in the "root" level and just set "styles" as active.

jasmussen commented:
> I think we should make this an experiment if we judge that the proposal is a temporarily worse experience than trunk. To be honest, I'm not entirely sure :P I feel like this is already better than trunk. But I also understand if you want to keep it "more hidden" until we iron out more things.

Oh right yes, the plugin is also a way to test. I'm super into _not_ hiding this behind an experiment, though if that's the case, we do need to nail down the interactivity, or lack of, in the frame itself. That plus the contextuality of the frame showing categories contextually to what you select on the left + porting over the revisions, resets, etc.

Another thing that's important to consider is: right now there's a use case, potentially common, of seeing your _site_ or another template, as you edit your global styles. Should there be a toggle, like an inverted version of this button, to let you see your main site template instead? 

![Screenshot 2024-09-25 at 12 05 23](https://github.com/user-attachments/assets/12b243fa-1331-4182-8b74-875e43bb19bc)

jorgefilipecosta commented:
So, we need to decide if this should be an experiment or something we can merge and then improve upon. What are the minimum viable changes that are required to merge an initial PR?

Would making the styles sidebar show all the other menu items with styles selected plus a button to show the home template be enough? For an initial version? It would then allow us to work in parallel with multiple follow-ups.

jorgefilipecosta committed:
Show other sidebar items on Styles sidebar

jorgefilipecosta committed:
refeactor sidebar current item

andrewserong commented:
This is looking very cool! Just pinging @ramonjd and @aaronrobertshaw for visibility as they're also looking at some Style Book work (though more focused on the body of the Style Book, whereas this is exposing the Styles section).

ramonjd commented:
I am just testing at end of my day. I think it's a great start to make global styles more accessible. Thank you!

>  Or does it select individual block types? If it's the latter, then it should navigate the the appropriate block in the 2nd column, just like it does in trunk.

> It would be cool if the stylebook focused or selects a block when you open its dedicated section in the sidebar.

I agree this would be a helpful feature to make editing block global styles easier. 

> We can have a dedicated section of the style book that is specific to "layout",

As part of the work going on in https://github.com/WordPress/gutenberg/issues/64707, color (and possibly) typography preset previews will also be rendered to the style book. Layout would be a nice additions.

What do folks think about the position of the popovers? I doesn't necessarily bother me personally, however for some reason my brain expects them to hover over the canvas. Seems more _feng shui_ for the popovers to point in the direction of the canvas. It's a small thing however right now.

<img width="300" alt="Screenshot 2024-09-26 at 4 34 44 pm" src="https://github.com/user-attachments/assets/f9589a91-feb9-48c7-83f1-d618ead2f4e2">

> What are the minimum viable changes that are required to merge an initial PR?

As others have mentioned, I reckon:

- Activate the click events for blocks to open block styles - the blocks have a selected state already so it seems like _something_ should happen
- Add the menu items (and then decide if they'll help or hinder) I personally think reset is enough for now. Revisions might complicate things conceptually, but that's just me.
- Have some way to open the editor, e.g., "a button to show the home template"


Later on:

- It might make sense to me to show the block category tabs to make things functionally consistent. Not sure. I think we're at the stage where we'll need to think about how to accommodate `n` registered categories in the style book as the tabs will be hidden in smaller screens.

youknowriad commented:
> What do folks think about the position of the popovers?

I agree it seems more natural on the right. I think it can be a follow-up though, I can see a "position" prop exposed on the GlobalStylesUI component.

youknowriad commented:
<img width="271" alt="Screenshot 2024-09-26 at 10 28 20 AM" src="https://github.com/user-attachments/assets/e9f82fc6-8e32-4765-b5bd-fd482d833bf2">

This message should be removed (or alternatively, we should try not navigating the sidebar entirely)

jorgefilipecosta committed:
button to toggle between stylebook and homepage

jorgefilipecosta committed:
Add path sync mechanism and sylebook block selection

jorgefilipecosta commented:
Thank you all for the reviews.
This PR passed by several changes to apply the suggested feedback:
- Style book now allows to select the blocks.
- The path contains the specific style section, e.g: the user can now link directly to typography or colors. This also includes everything we need to render different style books depending on the style section/path.
- We have now a button to toggle between the homepage and the style book view.
- We show all the menu items on the style section.
- The description of the styles section was changed.

The popovers should go to the right but I think that can be a simple follow-up.

There are many follow-ups and things we can do but I think it would be nice to merge this one as soon as possible as it open the door to many tasks we can work on in parallel/independent PR's.

youknowriad commented on packages/edit-site/src/components/global-styles/ui.js:
```diff
@@ -276,120 +277,158 @@ function GlobalStylesEditorCanvasContainerLink() {
 }
 
-function GlobalStylesUI() {
+function NavigationSync( { path: parentPath, onPathChange, children } ) {
```
It's unclear why is this a component. It should be just a hook.

jorgefilipecosta replied:
The reason why it is a component is to allow conditional application of the effect. A component can be applied conditionally while a hook can not.
```
const WrapperComponent = path && onPathChange ? NavigationSync : Fragment;
```

So for the uncontrolled case, we don't execute the effect at all. I think in this case a Component is a better choice, we can extract the effect as a hook anyway, having a component and a hook inside.

youknowriad replied:
I'm not sure, I still prefer a hook personally and just return early in the effect. I agree though that it's unfortunate that we can't use conditional hooks. I think component make more sense when we want to render something and not just return "children".

jorgefilipecosta replied:
Hi @youknowriad I tried to use the hook without a wrapper component but I noticed we need a wrapper component anyway because the hook calls need to be nested inside NavigatorProvider. So given that we are forced to have a wrapper I think the current solution is the preferred one.

youknowriad commented on packages/edit-site/src/components/global-styles/ui.js:
```diff
@@ -276,120 +277,158 @@ function GlobalStylesEditorCanvasContainerLink() {
+	const previousParentPath = usePrevious( parentPath );
+	const previousChildPath = usePrevious( childPath );
+	useEffect( () => {
+		console.log({
```
Debug code left

jorgefilipecosta replied:
Nice catch 😅

youknowriad commented on packages/edit-site/src/components/global-styles/ui.js:
```diff
@@ -276,120 +277,158 @@ function GlobalStylesEditorCanvasContainerLink() {
+	return children;
+}
+
+function GlobalStylesUI( { path, onPathChange } ) {
```
My understanding is that this change is making the `GlobalStylesUI` component controlled. Also by reading the code, it seems that it supports both controlled and uncontrolled modes. I think that's a good approach. Just trying to put it in words to clarify :)

jorgefilipecosta replied:
That's exactly it the component supports controlled and uncontrolled modes.

youknowriad commented on packages/edit-site/src/components/layout/router.js:
```diff
@@ -75,6 +75,31 @@ function useRedirectOldPaths() {
 
+const GLOBAL_STYLES_PATH_PREFIX = '/wp_global_styles';
+
+function GlobalStylesUIWrapper() {
```
For me we should move this component/folder outside of the "router.js" file, the router.js file should be just about "route registration" and not have a lot of logic/components.

jorgefilipecosta replied:
Good point I will do the change.

jorgefilipecosta committed:
apply code feedback

jorgefilipecosta commented:
Hi @youknowriad the feedback was applied.

jameskoster commented:
This looks like a promising start to me. A couple of high-level pieces of feedback:

* I'm not convinced that the frame should be interactive at all. Introducing this different interaction pattern feels pretty clunky to me. I'd lean towards disabling all interactivity (including the hover animation).
  * I could go either way about resizing – it doesn't seem super useful when viewing the stylebook?
* The "Styles" menu item should not be a drilldown, it should just open the styles panel on click.
  * The revisions button will need to move to the content frame, I think that's okay. Maybe we can mimic the current placement (IE in the panel header)
  * The stylebook toggle button will also need to move to the content frame. Again we can potentially copy the placement from the current Styles panel for now.
* Styles panel should adopt logic from data views for internal horizontal padding, IE utilise container queries. Example [here](https://github.com/WordPress/gutenberg/blob/3e0d294d59a60f0e87539a5465237262e239be38/packages/dataviews/src/components/dataviews/style.scss#L77). It would be good to codify this more generally at some point, likely as part of a dedicated `Page` component.

jorgefilipecosta committed:
no styles drill down; disable style books click;

jorgefilipecosta commented:
Hi @jameskoster,

I removed the frame interaction and resizing and made changes to the styles menu to prevent it from drilling down. Now it just opens a panel.

> Styles panel should adopt logic from data views for internal horizontal padding, IE utilise container queries. Example [here](https://github.com/WordPress/gutenberg/blob/3e0d294d59a60f0e87539a5465237262e239be38/packages/dataviews/src/components/dataviews/style.scss#L77). It would be good to codify this more generally at some point, likely as part of a dedicated Page component.

I'm a bit confused about what needs to be fixed. I currently don't see any horizontal scrolling on the styles panel. Could you possibly send a screenshot of the issue that needs to be addressed? Thank you a lot!

youknowriad commented on packages/edit-site/src/components/layout/router.js:
```diff
@@ -148,15 +150,24 @@ export default function useLayoutAreas() {
-				preview: <Editor />,
-				mobile: hasEditCanvasMode && <Editor />,
+				content: <GlobalStylesUIWrapper />,
+				preview: <Editor isPreviewOnly />,
```
For some reason, when I navigate to the "styles" menu, the frame doesn't animate anymore (you don't see it moving to the right), I'm pretty sure it used to animate earlier in the PR

jorgefilipecosta replied:
Hi @youknowriad, that was intentional following the feedback from @jameskoster "The "Styles" menu item should not be a drill-down, it should just open the styles panel on click.".
I think removing the drill-down removes the animation of the panel opening, I think we can not easily avoid having the drill down and also have the animation when the panel opens.

youknowriad replied:
> I think removing the drill-down removes the animation of the panel opening, I think we can not easily avoid having the drill down and also have the animation when the panel opens.

Why, how is the drill down related to the animation?

jorgefilipecosta replied:
I can easily re-add the animation by changing key from default to styles:
```
	if (path && path.startsWith('/wp_global_styles')) {
		return {
			key: 'styles',
```

The problem is that doing this doesn't just animate the style panel, but also animates the main "Design" area, which feels strange because there is no change in that part, only the styles item gets the blue background.

It seems that we can either animate the "Design" section and the style area, or we don't animate any of the areas. I don't see an option to not animate the style area but animate the styles panel.

youknowriad replied:
Maybe it's better to avoid animating based on the key, but let's not get blocked with that here, we can look at the animation separately.

jameskoster commented:
@jorgefilipecosta I'm referring to the internal padding of the Styles panel. Ideally it is aligned with Data Views. Here's a video to demonstrate how it works:

https://github.com/user-attachments/assets/ed2edb05-7c35-4210-84f6-b71026f163f5

When the container is wider than `430px` internal padding is `48px`, otherwise it is `24px`.

Currently the internal padding of the Styles panel is `16px` which produces an uncanny mismatch when you navigate from Pages to Styles. 

Like I said it would be good to codify this somehow, likely as a `Page` component, so that we don't have to apply it manually to every new section we add. I appreciate we can't do that here :) If we can't include the container query either than we should at least update the padding to `24px`.

Hope that makes sense!

jorgefilipecosta committed:
include padding logic

jorgefilipecosta commented:
Hi @jameskoster,
I Implemented the padding logic depending on container size as suggested. I could not codify at the page component devel because some UI have dividers and things like where we don't want the padding to apply.

jorgefilipecosta committed:
lint fixes

jameskoster commented:
Thanks @jorgefilipecosta. I see you created the `24px` overall padding by adding `12px` to `.edit-site-global-styles-sidebar__navigator-provider`. It works fine when the container is narrow, but when it's larger there's effectively `60px` padding (48 + 12). 

The spacing between the 'Styles' and 'Typography' headings here is a bit tight:

<img width="408" alt="Screenshot 2024-10-07 at 13 49 01" src="https://github.com/user-attachments/assets/c8135a46-c9af-4393-9cb7-ca71bc44fb6c">

I think we can add `12px` top padding to `.edit-site-global-styles-sidebar__navigator-provider`.

Small detail; the 'Styles' item in the main menu should not have a chevron icon as it doesn't drill down.

jorgefilipecosta committed:
feedback

jorgefilipecosta committed:
make padding equal to templates

jorgefilipecosta commented:
Hi @jameskoster I applied all your feedback.

> Thanks @jorgefilipecosta. I see you created the 24px overall padding by adding 12px to .edit-site-global-styles-sidebar__navigator-provider. It works fine when the container is narrow, but when it's larger there's effectively 60px padding (48 + 12).


I'm not sure how much padding we would want but I decreased what we had, it is now exactly equal to the template one I think that's what we want because of consistency.

jameskoster commented:
Thanks @jorgefilipecosta, I reckon we can begin gathering wider feedback now.

One small remaining issue I noticed is that panel borders do not span the full width of the container which looks a little odd. I wonder if we could just hide those in this context for now?

| Current | Hidden |
| --- | --- |
| <img width="423" alt="Screenshot 2024-10-08 at 15 55 39" src="https://github.com/user-attachments/assets/f530539c-de8b-4108-8fea-276ba2d51acf"> | <img width="426" alt="Screenshot 2024-10-08 at 15 57 14" src="https://github.com/user-attachments/assets/93761f45-7132-4535-821d-a1053c54948a"> |

jorgefilipecosta committed:
remove borders

jameskoster commented:
A couple of other small pieces (sorry for not catching these before, there's a lot to review here!):

* Hovering over the preview frame causes an animation, and clicking it reveals a focus ring. Ideally the preview frame is entirely inert (except for scrolling) when Stylebook is visible.
* If the site is visible it's possible to select blocks causing the toolbar to appear.
* Placement of the site/style book toggle feels off. I wonder if we place a menu in the Styles header?

<img width="1744" alt="menu" src="https://github.com/user-attachments/assets/c6a4907f-abb4-4706-a6e5-5e569f390910">

annezazu commented:
Tested this using Playground and didn't find an option to toggle between viewing site pages and the style book. I think that's necessary to have as, without it, it becomes hard to fully remove Styles from within the editor itself as it's the only way to see your site overall as you're making changes. Otherwise, I really like how this is pulled out at a higher level and think, at baseline, it immediately communicates the global nature much more concretely.

andrewserong commented:
> I think that's necessary to have as, without it, it becomes hard to fully remove Styles from within the editor itself as it's the only way to see your site overall as you're making changes

Is that the long-term objective? I'd assumed that we'd still want Styles to be within the editor as well, otherwise if someone wants to adjust some element of global styles while editing a template, they'll have to back out of editing and go all the way to the top again. Or is it confusing for it to be available in two places?

> Otherwise, I really like how this is pulled out at a higher level and think, at baseline, it immediately communicates the global nature much more concretely.

Agreed, I really like how this makes adjusting styles feel more immediate once you land in the site editor, and much less buried.

jameskoster commented:
@annezazu the switcher is in the dark sidebar. There's a design in https://github.com/WordPress/gutenberg/pull/65619#issuecomment-2401963293 to move it.

> Is that the long-term objective? [...] Or is it confusing for it to be available in two places?

I've seen a lot of feedback suggesting that folks often confuse the global styles sidebar with the block inspector.

It should be easy to access global styles from the site editor, but I don't think I've seen a design concept demonstrating what that flow should look like.

andrewserong commented:
Tiny update: I've chatted with a few folks who are interested in this PR and who've volunteered to help get this over the line while Jorge's away, so just adding a couple people as assignees on this one.

ramonjd committed:
Manual update of route files after https://github.com/WordPress/gutenberg/pull/66030

ramonjd committed:
Rename classname
Tweak paddings so that the scrollbar fits snug against the right hand side of the parent container

ramonjd commented:
I just rebased this — most of the manual changes were after https://github.com/WordPress/gutenberg/pull/66030 — and added a few updates. Hope I didn't break it. Working pretty well so far 🤔

ramonjd committed:
Fragment doesn't accept props
Scroll to block where a path is passed to Style book

ramonjd commented:
3cbc8cf scrolls to the block example (if it exists) when editing a block's styles:

https://github.com/user-attachments/assets/11305602-9913-46ec-8fee-24564b508628

ramonjd commented on packages/edit-site/src/components/style-book/index.js:
```diff
@@ -46,20 +53,44 @@ const {
+	if ( path && typeof path === 'string' && path.includes( '/blocks/' ) ) {
+		const block = decodeURIComponent( path.split( '/blocks/' )[ 1 ] );
+		return {
+			block,
```
Returns an object to leave open the possibility of deep linking to `category > block` later.

ramonjd commented on packages/edit-site/src/components/style-book/index.js:
```diff
@@ -46,20 +53,44 @@ const {
+	const element = frameWindow.document.getElementById( anchor );
+	if ( element ) {
+		element.scrollIntoView( {
+			behavior: 'smooth',
```
Or we just snap to the anchor with no scroll. Either or.

ramonjd commented on packages/edit-site/src/components/style-book/index.js:
```diff
@@ -46,20 +53,44 @@ const {
 	isSelected,
-	onClick,
-	onSelect,
+	onClick = noop,
```
Adding `noop` to avoid having to pass them in the props

ramonjd commented on packages/edit-site/src/components/style-book/index.js:
```diff
@@ -210,8 +246,17 @@ const StyleBookBody = ( {
+	const handleLoad = () => setHasIframeLoaded( true );
+	useLayoutEffect( () => {
+		if ( goTo?.block && hasIframeLoaded && iframeRef?.current ) {
+			scrollToSection( `example-${ goTo?.block }`, iframeRef?.current );
```
On page refreshes this might not be 100% accurate. I assume because some of the iframe assets are still being rendered and therefore the document height value is inaccurate.

It's an edge case - here we're mainly interested in arriving at the style book via the style menu. The iframe will have loaded by the time a user navigates to the block styles I expect.

tellthemachines replied:
It does seem to be a minor issue. The main scenarios would probably be bookmarking a specific block, sharing the link with someone or opening in another tab/browser. Not sure how common these are though.
That said, if it's a block far down the list and the screen isn't huge (say regular laptop size), there's a good chance the selected block won't even appear on screen 😅 

I guess if we were displaying the tabbed interface here with the blocks divided into sections, no section would be long enough for that to be too noticeable.

tellthemachines reviewed (commented):
Overall it makes a lot of sense to move the global styles sidebar to the left. I'm not sure about adding the whole thing to the second sidebar though. 

How difficult might it be to make the root level global styles menu appear in the black sidebar instead? I think that would be an important improvement to this piece of work:

* It's more consistent with how the Pages section opens up to a page-specific menu;
* It would avoid having a large slice of essentially wasted space (because the root menu on that black sidebar isn't really useful to have when editing styles) which can be quite annoying on a small screen:
<img width="1288" alt="Screenshot 2024-10-23 at 4 06 26 pm" src="https://github.com/user-attachments/assets/884d85b3-262d-4c68-abe7-e376e0fc3197">
* It would allow us to preserve the current experience of Styles section opening up directly onto the theme style variations and colour palettes (we could make "Browse styles" be the default selection in the white sidebar when it opens).

What do you folks think?

jasmussen commented:
Created #66376 as a way to better track this work and connect dots.

ramonjd committed:
Added some function docs

ramonjd committed:
Remove noop because the presence of onClose determines whether a close button appears in the style book

ramonjd committed:
Scroll to colors

ramonjd committed:
Move control of style book to global styles panel.

andrewserong commented:
I like the idea of putting the toggle to show / hide the style book at the top of the Styles panel 👍 

https://github.com/user-attachments/assets/034a4d3c-5e02-4034-b9a9-bec3199591a7

In terms of making it more discoverable, I also liked the example from @jameskoster in https://github.com/WordPress/gutenberg/issues/66376#issuecomment-2442079891 of side-by-side toggle buttons:

<img width="847" alt="image" src="https://github.com/user-attachments/assets/d65b77a2-577d-43f8-a2ce-da681d0da6bc">

And I see there's discussion about potentially using text labels, too, which could be clearer (https://github.com/WordPress/gutenberg/issues/66376#issuecomment-2445502281). Whenever I see the template / layout icon I think it's descriptive rather than something that can be clicked 🤔

---

Separately to the icon question, when we're previewing the site in the right hand frame, I wonder if clicking it should switch us in to editing within the site editor proper? When I clicked on the template preview I expected to be able to go into template editing. That said, I'm aware earlier in this thread (https://github.com/WordPress/gutenberg/pull/65619#issuecomment-2379487543) there was a comment that the frame shouldn't be interactive — I just wasn't too sure if that was specifically about the Style Book or if it was meaning no interactivity there altogether.

---

In any case, thanks for keeping this PR current @ramonjd and for pushing it along!

ramonjd commented:
> if we don't make the right-hand side interactive to get to the template editor, users might feel stuck.

Yes, I wanted to click on the canvas preview to open the editor too, given that's the current behaviour of the other pages in the site editor nav.

I can understand that, with the styles panel open, the user is currently editing global styles, which is separate from template editing so clicking on the preview would interrupt that. And where would it lead to anyway? If users are viewing a template, clicking on it could open the editor, but I'm not sure about what should happen where the style book is loaded. 

I'm in favour of making things as clear as possible - in this case, as few exceptions to a user's expectations as possible. So, if clicking on the preview should do nothing, then have an "open in editor" button on this screen, which would open the editor in edit mode and load the template/style book respectively.

> I like the idea of putting the toggle to show / hide the style book at the top of the Styles panel 

👍🏻 

Putting the question of whether it should be text or a toggle control aside, I think this is a better place for it because it's closer to the thing that's changing. That's just a personal preference. Happy to roll back and change things up. 👍🏻 

On a related note: part of the greater effort is to have the style panel available in [classic themes](https://github.com/WordPress/gutenberg/issues/41119), so I doubt the dark menu (and therefore any stylebook toggle button) would be there anyway (?).

ntsekouras commented on packages/edit-site/src/components/resizable-frame/index.js:
```diff
@@ -82,6 +82,7 @@ function ResizableFrame( {
 	isOversized,
 	setIsOversized,
 	isReady,
+	hasResize = true,
```
That's a weird prop in ResizableFrame component. Do we need to use it if we don't want to resize?

ramonjd replied:
Yeah, I couldn't work it out either. I'm assuming we don't want this quite yet! 😄 


https://github.com/user-attachments/assets/0d537490-bf86-461c-9245-6c174fda8daf

ntsekouras commented on packages/edit-site/src/components/sidebar-global-styles-wrapper/index.js:
```diff
@@ -0,0 +1,94 @@
+					<>
+						{ ! isMobileViewport && (
+							<SidebarButton
+								icon={ isStyleBookOpened ? layout : seen }
```
Do you think changing icons is a good idea? I found it a bit confusing and I think there are a11y concerns about doing this.

ramonjd replied:
Not really. The last commit was just moving them into the global styles panel. I'll try something text-based and folks can pile on with the opinions.

ntsekouras reviewed (commented):
Thank you for working on this! I tested it a bit and feels quite good in general.

Some thoughts are:
1. When we preview the template I feel like it shouldn't be clickable and not have styles to indicate it is (like the cursor)
2. If we click on a block in style book, should it open the respective panel like in trunk?

jameskoster commented:
@ramonjd there was a design suggestion to use a menu for the stylebook toggle [here](https://github.com/WordPress/gutenberg/pull/65619#issuecomment-2401963293). Worth a try?

> If we click on a block in style book, should it open the respective panel like in trunk?

I would lean towards keeping the Preview Frame inert (aside from scrolling) in this PR. It has a very singular purpose currently (it's always one big button that opens the editor). Additional interactions should be carefully considered in the wider context of the site editor layout pieces.

ramonjd committed:
Removing interactivity on the preview resizable editor for the styles preview
Ensuring navigation works when clicking on a style book element

ramonjd commented:
> When we preview the template I feel like it shouldn't be clickable and not have styles to indicate it is (like the cursor)
> If we click on a block in style book, should it open the respective panel like in trunk?

Both good points. Done ✅

ramonjd commented:
> there was a design suggestion to use a menu for the stylebook toggle https://github.com/WordPress/gutenberg/pull/65619#issuecomment-2401963293. Worth a try?

Oh, just saw this. Yes, worth a try. Thank you!

ramonjd committed:
Remove icon in favour of icon + dropdown menu

ramonjd commented:
> there was a design suggestion to use a menu for the stylebook toggle https://github.com/WordPress/gutenberg/pull/65619#issuecomment-2401963293. Worth a try?

I implemented it in this PR, but left the check icon in the default position for `<MenuItem />`, which is `"right"`. 

`"left"` doesn't align with other items' texts.

| Default (right) | Left  |
|--------|--------|
| <img width="653" alt="Screenshot 2024-10-31 at 3 51 58 pm" src="https://github.com/user-attachments/assets/6340f9be-849c-4761-a239-a299b0d4cff2"> | <img width="287" alt="Screenshot 2024-10-31 at 3 56 23 pm" src="https://github.com/user-attachments/assets/4eef2e0b-932b-47ca-bb39-ae2e6557196d"> |

andrewserong commented:
I'm liking the dropdown! The descriptions in the dropdown help give clarity, too.

I still feel like it's easy to become a bit stuck and unable to get into editing the home page template now, as if you're already in the Styles screen, you need to know to click Templates in the sidebar and then select a template to go edit it. Is it worth adding an Edit icon in the header next to the dropdown a bit like what we currently have on trunk?

<img width="290" alt="image" src="https://github.com/user-attachments/assets/88362af0-ae2a-48db-ba36-7b2655781146">

I'm sure there'd be other ways to go about it, too (addressing the change to discoverability re: getting into the site editor proper), so just an idea!

jameskoster commented:
@ramonjd can we use the [Menu](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-menu--docs) component, and I suppose those should be `RadioItem`s? This is the newer menu component that we're planning to [migrate instances of `DropdownMenu` to](https://github.com/WordPress/gutenberg/issues/61094).

> I still feel like it's easy to become a bit stuck and unable to get into editing the home page template now

I don't mind the Preview Frame being clickable when previewing the _site_. It should only be inert when Style Book is active.

andrewserong commented:
> I don't mind the Preview Frame being clickable when previewing the site. It should only be inert when Style Book is active.

Perfect. That would address my concern for now! 👍

ramonjd committed:
Revert `isStaticPreview`
Use Menu component for dropdown

update copy

ramonjd commented:
> can we use the [Menu](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-menu--docs) component, and I suppose those should be RadioItems?
> I don't mind the Preview Frame being clickable when previewing the site. It should only be inert when Style Book is active.

Done and done.

https://github.com/user-attachments/assets/5088a641-4026-49c5-b0fc-943a222cc86d

A couple of observations:

1. The Menu component radio items don't yet allow custom checked icons. Current is a default circle. This can be done in another PR pretty easily though.
2. To prevent jank, the button has to have static text. Not such a bad thing, but I'm open to suggestions which are better than my contribution (`'Preview'`)

Otherwise LGTM

andrewserong commented:
> Done and done.

That's looking and feeling good to me! I like the dropdown and text-based preview button, and now that the site preview is clickable, I'm no longer feeling trapped in the Styles section 👍 😄

Let's see what folks think!

ramonjd commented:
> The Menu component radio items don't yet allow custom checked icons. Current is a default circle. This can be done in another PR pretty easily though.

- https://github.com/WordPress/gutenberg/pull/66666

aaronrobertshaw commented:
> The Menu component radio items don't yet allow custom checked icons. Current is a default circle. This can be done in another PR pretty easily though.

This is just my opinion but I associate `check` symbols with checkboxes and multi-selection. 

The options in this menu are an either-or proposition. So making them radio items makes a lot of sense to me. Displaying a `check` icon beside the selected option though leaves open the possibility for the user to expect to be able to choose both at the same time. 

It might still be desirable for the `Menu` component to accept custom checked prefixes in case consumers want to render a different kind of radio check 🤷

jasmussen commented:
This is a great PR, it feels very close to landing. To help unstick this, I created this mockup:

<img width="1840" alt="i2" src="https://github.com/user-attachments/assets/66822e45-28b3-4e9b-8732-e263a1654525">

The PR here is already virtually there, it's just a matter of updating the style of the preview button. Since we use tertiary blue link buttons in similar contexts, I'd go with such a one, here, as well, in its `compact` size. 

What's missing?

jasmussen commented:
Shared an update here too: https://github.com/WordPress/gutenberg/issues/66376#issuecomment-2451605902

jameskoster commented:
+1 what @aaronrobertshaw said. Apologies for sharing a misleading design above. We should indeed use the vanilla `MenuRadioItem` design.

richtabor commented:
Here's the current implementation for reference: 

https://github.com/user-attachments/assets/21508b0d-7014-4d10-b7f4-da3ffea69c47

richtabor commented:
One big issue of this current flow is that it's next to impossible to get back to the homepage of the site editor, unless you first change the "Preview" button state to "Site" to effectively click into your site. Otherwise, there's no way to get back to the top-level entry point of the site editor.

One solution would be to default to the "Site" instead of "Style book", so that at least you don't get stuck unless you first switch to "Style book". 

https://github.com/user-attachments/assets/8d0ab25c-7d71-40ea-8191-c46c0fad6061

richtabor commented:
> One solution would be to default to the "Site" instead of "Style book", so that at least you don't get stuck unless you first switch to "Style book".

I propose we default to the site rather than style book, to reduce this problem and be able to push this forward.

ramonjd commented:
> Displaying a check icon beside the selected option though leaves open the possibility for the user to expect to be able to choose both at the same time.

I agree. Despite the cool PR number, I closed https://github.com/WordPress/gutenberg/pull/66666 as well. 👍🏻 

> I don't like how the sidebar nav "resets" when selecting "Styles".

> One solution would be to default to the "Site" instead of "Style book", so that at least you don't get stuck unless you first switch to "Style book".

I agree because:

1. We can move on this PR faster, and move detailed conversation about style book and other interaction to new PR.
2. It doesn't break the current paradigm whereby users click on the canvas to edit.

ramonjd committed:
Remove stylebook
Disable animation for style view

ramonjd committed:
Revert changes to stylebook since their no longer required after 6c8a23f

ramonjd commented:
> I propose we default to the site rather than style book, to reduce this problem and be able to push this forward.

I've done this for now.

Here's the updated state:

https://github.com/user-attachments/assets/9a454759-ff7e-4289-887e-baef94ff3ed9

Proposed next steps:

1. Create a new PR to reintegrate the Style Book into the style view and discuss how to implement a consistent navigation

One thing to note is that trunk still has a stylebook and edit icons:

<img width="266" alt="Screenshot 2024-11-02 at 3 11 05 pm" src="https://github.com/user-attachments/assets/0d8af806-527c-4d24-9a81-514060e4aa59">

The other option for this PR is to keep the site/stylebook toggle, and allow canvas clicking to navigate to the editor for both.

andrewserong commented:
> I propose we default to the site rather than style book, to reduce this problem and be able to push this forward.
> I've done this for now.

I read the feedback as switching what we default to, to being the editor / template preview, rather than removing the Style Book toggle altogether?

That said, I agree if it helps land this PR sooner, we can add the toggle back in, in a follow-up 👍

> I don't like how the sidebar nav "resets" when selecting "Styles".

That fix is feeling nice to me!

ramonjd commented:
> I read the feedback as switching what we default to, to being the editor / template preview, rather than removing the Style Book toggle altogether?

Oh yeah, I misread. Thanks for pulling me up on that @andrewserong 

I can revert. 

I also perceived things were slowing down around how a user should switch between the two views (site and style book), as well as the copy, so that was also my motivation to move the conversation elsewhere.

Whatever folks think is better.

ramonjd committed:
Default to site, and reinstate style book

ramonjd commented:
> I read the feedback as switching what we default to, to being the editor / template preview, rather than removing the Style Book toggle altogether?

Reverted in [8902fa3](https://github.com/WordPress/gutenberg/pull/65619/commits/8902fa396d32a6733365c889673dd74e7ada31a9)

aaronrobertshaw commented on packages/edit-site/src/components/sidebar-global-styles-wrapper/index.js:
```diff
@@ -0,0 +1,145 @@
+	return (
+		<Menu
+			trigger={
+				<Button __next40pxDefaultSize variant="secondary">
```
```suggestion
				<Button __next40pxDefaultSize variant="tertiary" size="compact">
```

Looks like we're missing one of the final tweaks proposed in https://github.com/WordPress/gutenberg/pull/65619#issuecomment-2451585626.

> The PR here is already virtually there, it's just a matter of updating the style of the preview button. Since we use tertiary blue link buttons in similar contexts, I'd go with such a one, here, as well, in its compact size.

aaronrobertshaw commented on packages/edit-site/src/components/sidebar-global-styles-wrapper/index.js:
```diff
@@ -0,0 +1,145 @@
+			<Menu.RadioItem
+				value
+				checked={ isStyleBookOpened }
+				name="radio-controlled"
```
While we're tweaking this dropdown, would it be worth making the radio group name meaningful to this context?

aaronrobertshaw commented on packages/edit-site/src/components/sidebar-navigation-item/style.scss:
```diff
@@ -28,6 +28,8 @@
 	&.with-suffix {
 		padding-right: $grid-unit-20;
 	}
+
+
```
Tiniest of nits: Do we need the extra whitespace here?

ramonjd commented on packages/edit-site/src/components/sidebar-global-styles-wrapper/index.js:
```diff
@@ -0,0 +1,145 @@
+	const { params } = useLocation();
+	const history = useHistory();
+	const { canvas = 'view' } = params;
+	const [ isStyleBookOpened, setIsStyleBookOpened ] = useState( false );
```
The style book view is off by default.

andrewserong commented:
> Reverted in [8902fa3](https://github.com/WordPress/gutenberg/pull/65619/commits/8902fa396d32a6733365c889673dd74e7ada31a9)

Looking good to me!

https://github.com/user-attachments/assets/1d694e1d-1374-4dfc-9096-fa7310080feb

aaronrobertshaw reviewed (commented):
Great effort across the board on this one, everyone. Thanks for continuing to drive it forward @ramonjd.

It is testing well for me and appears to address nearly all the feedback received so far. I've left a couple of minor nits as inline comments below.

> I read the feedback as switching what we default to, to being the editor / template preview, rather than removing the Style Book toggle altogether?

This was my read as well. Thanks for reverting the other change there.

Defaulting to the site preview does appear to make it harder to fall into the trap of not easily getting back to the site preview and entering the editor. Good enough to land this PR with that approach, I think.

In case it helps, here's a small diff with suggested tweaks from the comments below.

<details>
<summary>Quick diff, feel free to take it or leave it 🤷 </summary>

```diff
diff --git a/packages/edit-site/src/components/sidebar-global-styles-wrapper/index.js b/packages/edit-site/src/components/sidebar-global-styles-wrapper/index.js
index 356385a10b..afa9f489dd 100644
--- a/packages/edit-site/src/components/sidebar-global-styles-wrapper/index.js
+++ b/packages/edit-site/src/components/sidebar-global-styles-wrapper/index.js
@@ -30,7 +30,7 @@ const GlobalStylesPageActions = ( {
 	return (
 		<Menu
 			trigger={
-				<Button __next40pxDefaultSize variant="secondary">
+				<Button __next40pxDefaultSize variant="tertiary" size="compact">
 					{ __( 'Preview' ) }
 				</Button>
 			}
@@ -38,7 +38,7 @@ const GlobalStylesPageActions = ( {
 			<Menu.RadioItem
 				value
 				checked={ isStyleBookOpened }
-				name="radio-controlled"
+				name="styles-preview-actions"
 				onChange={ () => setIsStyleBookOpened( true ) }
 				defaultChecked
 			>
@@ -50,7 +50,7 @@ const GlobalStylesPageActions = ( {
 			<Menu.RadioItem
 				value={ false }
 				checked={ ! isStyleBookOpened }
-				name="radio-controlled"
+				name="styles-preview-actions"
 				onChange={ () => setIsStyleBookOpened( false ) }
 			>
 				<Menu.ItemLabel>{ __( 'Site' ) }</Menu.ItemLabel>
diff --git a/packages/edit-site/src/components/sidebar-navigation-item/style.scss b/packages/edit-site/src/components/sidebar-navigation-item/style.scss
index f9f3a01868..202de53000 100644
--- a/packages/edit-site/src/components/sidebar-navigation-item/style.scss
+++ b/packages/edit-site/src/components/sidebar-navigation-item/style.scss
@@ -28,8 +28,6 @@
 	&.with-suffix {
 		padding-right: $grid-unit-20;
 	}
-
-
 }
 
 .edit-site-sidebar-navigation-screen__content .block-editor-list-view-block-select-button {

```
</details>

ramonjd committed:
Implementing feedback

ramonjd commented:
Thanks again for testing again folks. @aaronrobertshaw great suggestions. Done all round.

aaronrobertshaw reviewed (approved):
Lightning fast turnaround there! ⚡ 

I've given this a final smoke test and all the feedback not earmarked for follow-ups has been addressed. In my view, we should merge this one and iterate from there.

LGTM 🚢

jasmussen commented:
I'd make the preview button a tertiary button, compact size. Otherwise, congrats for bringing this forward!

aaronrobertshaw commented:
> I'd make the preview button a tertiary button, compact size. Otherwise, congrats for bringing this forward!

This was [addressed](https://github.com/WordPress/gutenberg/pull/65619/commits/4d5a7b5f2f4d65b96ffba1f6cc77bcc0d3c2f7c2#diff-ad9c0cee6c40faaa298bc050e1e7fe39614a79a95069b111fe32ca2bd6b19e16R33) in https://github.com/WordPress/gutenberg/pull/65619/commits/4d5a7b5f2f4d65b96ffba1f6cc77bcc0d3c2f7c2 after my earlier review.

jasmussen commented:
Fantastic, thank you!

jorgefilipecosta commented:
Thank you all for bringing this to the finish line 🥳
